### Import the necessary libraries

In [1]:
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from helpers.utils import load_data

### Load the datasets

In [3]:
traffic, stations, mappings = load_data()

### Data Cleaning

Let's check if there are any duplicate rows in traffic dataset

In [5]:
duplicates = traffic[traffic.duplicated()]
print(f"Traffic dataset contains {len(duplicates)} duplicate rows")
duplicates.head()

Traffic dataset contains 750678 duplicate rows


date  day_of_data  day_of_week  direction_of_travel  \
9440   2015-11-02            2            2                    1   
12434  2015-01-21           21            4                    5   
14835  2015-08-09            9            1                    7   
14888  2015-04-25           25            7                    1   
16694  2015-04-06            6            2                    3   

      fips_state_code functional_classification  month_of_data station_id  \
9440               54                        3R             11     000016   
12434              29                        2U              1     004913   
14835              54                        4U              8     000045   
14888              13                        3U              4     000222   
16694              39                        1R              4     000155   

       traffic_volume_counted_after_0000_to_0100  \
9440                                          25   
12434                                         49   
14835                                         23   
14888                                         27   
16694                                        184   

       traffic_volume_counted_after_0100_to_0200  ...  \
9440                                          23  ...   
12434                                         26  ...   
14835                                         20  ...   
14888                                         20  ...   
16694                                        121  ...   

       traffic_volume_counted_after_1400_to_1500  \
9440                                         136   
12434                                        351   
14835                                        260   
14888                                        252   
16694                                        893   

       traffic_volume_counted_after_1500_to_1600  \
9440                                         157   
12434                                        432   
14835                                        300   
14888                                        258   
16694                                        923   

       traffic_volume_counted_after_1600_to_1700  \
9440                                         176   
12434                                        887   
14835                                        200   
14888                                        265   
16694                                        932   

       traffic_volume_counted_after_1700_to_1800  \
9440                                         238   
12434                                       1407   
14835                                        187   
14888                                        258   
16694                                        898   

       traffic_volume_counted_after_1800_to_1900  \
9440                                         210   
12434                                       1546   
14835                                        219   
14888                                        213   
16694                                        731   

       traffic_volume_counted_after_1900_to_2000  \
9440                                         130   
12434                                       1015   
14835                                        159   
14888                                        172   
16694                                        573   

       traffic_volume_counted_after_2000_to_2100  \
9440                                          77   
12434                                        377   
14835                                        136   
14888                                        122   
16694                                        476   

       traffic_volume_counted_after_2100_to_2200  \
9440                                          53   
12434                                        243   
14835                                        108   
14888                                         93   
16694                             

Given that it is very unlikely for a data row to be exactly identical to another for all of its 32 columns (including the traffic volume for all the time intervals). It is highly likely that these data rows are genuine cases of duplicate entries, arising from errors made from the data collection/collation process. As such, let's drop these duplicate rows from our subsequent analysis.

In [6]:
traffic = traffic.drop_duplicates(ignore_index = True)
print(f"Number of rows left in unduplicated traffic dataset: {len(traffic)}")

Number of rows left in unduplicated traffic dataset: 6389713


Let's also check for invalid values in the traffic volume columns (i.e. negative values)

In [7]:
start_column = "traffic_volume_counted_after_0000_to_0100"
end_column = "traffic_volume_counted_after_2300_to_2400"

invalid = traffic[(traffic.loc[:, start_column: end_column] < 0).any(axis=1)]
print(f"There are {len(invalid)} rows containing negative values for traffic volume")
invalid.head()

There are 851 rows containing negative values for traffic volume


date  day_of_data  day_of_week  direction_of_travel  \
22298  2015-03-08            8            1                    3   
32468  2015-03-08            8            1                    3   
43542  2015-03-08            8            1                    5   
45564  2015-03-08            8            1                    3   
55491  2015-03-08            8            1                    3   

      fips_state_code functional_classification  month_of_data station_id  \
22298              51                        5R              3     130018   
32468              51                        4R              3     150326   
43542              51                        1R              3     781229   
45564              51                        1U              3     040766   
55491              51                        1U              3     090122   

       traffic_volume_counted_after_0000_to_0100  \
22298                                         10   
32468                                         14   
43542                                          8   
45564                                        667   
55491                                        427   

       traffic_volume_counted_after_0100_to_0200  ...  \
22298                                          7  ...   
32468                                          5  ...   
43542                                         -1  ...   
45564                                         -1  ...   
55491                                         -1  ...   

       traffic_volume_counted_after_1400_to_1500  \
22298                                        247   
32468                                         96   
43542                                         40   
45564                                       2733   
55491                                        689   

       traffic_volume_counted_after_1500_to_1600  \
22298                                        270   
32468                                         78   
43542                                         50   
45564                                       2899   
55491                                        682   

       traffic_volume_counted_after_1600_to_1700  \
22298                                        270   
32468                                         58   
43542                                         44   
45564                                       2886   
55491                                        704   

       traffic_volume_counted_after_1700_to_1800  \
22298                                        238   
32468                                         71   
43542                                         36   
45564                                       2759   
55491                                        678   

       traffic_volume_counted_after_1800_to_1900  \
22298                                        162   
32468                                         75   
43542                                         22   
45564                                       2709   
55491                                        586   

       traffic_volume_counted_after_1900_to_2000  \
22298                                        113   
32468                                         63   
43542                                         26   
45564                                       2161   
55491                                        530   

       traffic_volume_counted_after_2000_to_2100  \
22298                                         49   
32468                                         60   
43542                                         28   
45564                                       1698   
55491                                        451   

       traffic_volume_counted_after_2100_to_2200  \
22298                                         43   
32468                                         21   
43542                                         18   
45564                                       1295   
55491                             

Let's inspect what the negative values are and their respective counts

In [8]:
invalid_traffic_volume = invalid.loc[:, start_column: end_column].values

negative_values, counts = np.unique(
    invalid_traffic_volume[invalid_traffic_volume < 0],
    return_counts = True
)

print("Value: Count")
for value, count in zip(negative_values, counts):
    print(f"{str(value).rjust(5, ' ')}: {count}")

Value: Count
-3061: 1
   -1: 1448


In most cases (1448 out of 1449), -1 was the negative value. These negative values could be because the sensors were down for that particular time interval. Let's assume that the negative values were indeed due to the down sensors. In such a scenario, no values would be recorded for traffic volume (i.e. traffic volume = 0). As such, let's replace the negative values with 0.

In [9]:
replacements = {col: {value: 0 for value in negative_values} for col in traffic.columns[-24:]}
traffic = traffic.replace(replacements)

### Caveats

Let's check that in general, the traffic dataset does contain 365 days (a year) of data

In [10]:
print(f"Number of unique dates in traffic dataset: {len(traffic.date.unique())}")

Number of unique dates in traffic dataset: 365


For the purpose of subsequent analysis in caveats, let's define a unique station based on its fips_state_code, station_id, direction_of_travel and function_classification. What this means is that so long as two stations have different fips_state_code, station_id, direction_of_travel or function_classification, they are considered as two unique stations.

Let's check if evey unique station has 365 days (a year) worth of data

In [11]:
columns = ["date", "fips_state_code", "station_id", "functional_classification", "direction_of_travel"]
group_columns = ["fips_state_code", "station_id", "direction_of_travel", "functional_classification"]

num_unique_dates = traffic[columns] \
.groupby(group_columns) \
.agg("nunique") \
.reset_index(drop=False) \
.rename(columns = {"date": "num_unique_dates"})

num_unique_dates

fips_state_code station_id  direction_of_travel  \
0                  01     000002                    3   
1                  01     000002                    7   
2                  01     000004                    3   
3                  01     000004                    7   
4                  01     000006                    1   
...               ...        ...                  ...   
12369              56     0169NE                    5   
12370              56     0169NW                    1   
12371              56     0169NW                    5   
12372              56     RP0369                    1   
12373              56     RP0370                    5   

      functional_classification  num_unique_dates  
0                            3R               325  
1                            3R               324  
2                            1R               330  
3                            1R               330  
4                            4R               287  
...                         ...               ...  
12369                        3R               347  
12370                        4R               347  
12371                        4R               347  
12372                        1U               359  
12373                        1U               359  

[12374 rows x 5 columns]

From the table above, it seems that not all the unique stations have 365 days of recorded data. Let's take a look at the box plot for num_unique_dates.

In [12]:
num_unique_dates[["num_unique_dates"]].describe()

num_unique_dates
count      12374.000000
mean         301.745919
std           87.625571
min            1.000000
25%          286.000000
50%          342.000000
75%          360.000000
max          365.000000

In [42]:
# It appears that not every station has all 365 days of its data recorded
# Let's take a look at the summary statistics for num_unique_dates
num_unique_dates[["num_unique_dates"]].describe()

num_unique_dates
count      12374.000000
mean         301.745919
std           87.625571
min            1.000000
25%          286.000000
50%          342.000000
75%          360.000000
max          365.000000

In [43]:
# Let's sort the data by num_unique_dates
num_unique_dates.sort_values("num_unique_dates").head(100)

fips_state_code station_id  direction_of_travel  \
4766               22     000057                    5   
4765               22     000057                    1   
1347               06     043340                    3   
11584              54     000090                    3   
1511               06     075020                    3   
...               ...        ...                  ...   
839                04     101249                    1   
838                04     101247                    5   
130                01     000156                    1   
9162               47     000235                    3   
2846               13     000112                    7   

      functional_classification  num_unique_dates  
4766                         4R                 1  
4765                         4R                 1  
1347                         1U                 1  
11584                        1U                 1  
1511                         1U                 1  
...                         ...               ...  
839                          2U                24  
838                          2U                24  
130                          1U                25  
9162                         1U                25  
2846                         2U                25  

[100 rows x 5 columns]

In [44]:
# Let's find the number of unique stations per state
# For our analysis, a unique station is defined based on its station_id, direction_of_travel and functional_classification
# i.e. two stations with the same station_id but different direction_of_travel is considered as 2 unique stations
num_stations = num_unique_dates.groupby(
    ["fips_state_code", "station_id", "direction_of_travel", "functional_classification"]) \
    .agg("count") \
    .reset_index() \
    .rename(columns = {"num_unique_dates": "num_unique_stations"})

num_unique_stations = num_stations[["fips_state_code", "num_unique_stations"]] \
                      .groupby("fips_state_code") \
                      .sum() \
                      .reset_index()
num_unique_stations

fips_state_code  num_unique_stations
0               01                  296
1               02                  210
2               04                  526
3               05                  115
4               06                  570
5               08                  212
6               09                   75
7               10                  168
8               11                    4
9               12                  545
10              13                  446
11              15                  152
12              16                  423
13              17                  205
14              18                  171
15              19                  201
16              20                  196
17              21                  162
18              22                  120
19              23                  129
20              24                  150
21              25                  103
22              26                  245
23              27                  177
24              28                  194
25              29                  168
26              30                  197
27              31                   99
28              32                  196
29              33                  195
30              34                   75
31              35                  228
32              36                  345
33              37                  134
34              38                  128
35              39                  352
36              40                  180
37              41                  350
38              42                  180
39              44                  184
40              45                  131
41              46                  131
42              47                  114
43              48                  528
44              49                  220
45              50                  230
46              51                  961
47              53                  327
48              54                  141
49              55                  447
50              56                  338

In [45]:
# There is a large spread in terms of the number of unique stations in each state as seen from below
num_unique_stations.describe()

num_unique_stations
count            51.000000
mean            242.627451
std             166.828650
min               4.000000
25%             137.500000
50%             195.000000
75%             311.500000
max             961.000000

In [62]:
# Let's check if there is a correlation between the annual traffic volume and the number of unique stations in each state
# Intuitively, if a state has more unique stations, its annual traffic volume should also be higher, since it has more
# sensors that are tracking the traffic volume
# If our hypothesis is true, then annual traffic volume by state is not an accurate metric to measure how heavy the traffic is
# in a state, as the high traffic volume might just be because of the large number of sensors used for tracking
# In such a scenario, annual traffic volume normalized by the number of unique stations might be a more accurate metric to use.
start_column = "traffic_volume_counted_after_0000_to_0100"
end_column = "traffic_volume_counted_after_2300_to_2400"
traffic["total_daily_traffic_volume"] = traffic \
                                        .loc[:, start_column: end_column] \
                                        .sum(axis=1)

traffic_state = traffic[["fips_state_code", "total_daily_traffic_volume"]] \
                .groupby("fips_state_code") \
                .sum() \
                .reset_index() \
                .rename(columns = {"total_daily_traffic_volume": "annual_traffic_volume"})
traffic_state

fips_state_code  annual_traffic_volume
0               01             1461509905
1               02              318138924
2               04             5600393206
3               05              278436058
4               06             6670057776
5               08              964145461
6               09              500556105
7               10              534104309
8               11               59205437
9               12             4412694931
10              13             3094751035
11              15             1605036832
12              16             1134837654
13              17              604663105
14              18              332236497
15              19              510570289
16              20              340832204
17              21              597736449
18              22              263211358
19              23              195938846
20              24             1399825682
21              25              795943829
22              26             2308504217
23              27              828942796
24              28              619803351
25              29              849276076
26              30              255608184
27              31              329863440
28              32             1037072029
29              33             1108253743
30              34              549456912
31              35              384977104
32              36             1560213283
33              37              466053858
34              38               80318777
35              39             2631835767
36              40             1090108811
37              41             1443103445
38              42              534070530
39              44             2441569023
40              45             1511178572
41              46              147911499
42              47              276729299
43              48             2791691757
44              49             1191135280
45              50              197420518
46              51             9056241744
47              53             2702688901
48              54              151767797
49              55             1872783834
50              56              279254443

In [71]:
# Merge num_unique_stations with traffic_state
merge_station_state = num_unique_stations.merge(traffic_state, on="fips_state_code")
fig = px.scatter(merge_station_state, x="num_unique_stations", y="annual_traffic_volume",
           color="fips_state_code", trendline="ols", trendline_scope="overall",
           hover_data=["fips_state_code"], trendline_color_override="black",
           labels={"num_unique_stations": "No of Unique Stations",
                   "annual_traffic_volume": "Annual Traffic Volume"})
fig

In [81]:
# Based on the graph above and the R^2 value = 0.769 (can be found by hovering over the trendline in the graph), we can  conclude
# that there is indeed a strong positive correlation between num_unique_stations and the annual traffic volume
# As such, it might be prudent to measure how heavy traffic is in a state using the metric: annual traffic volume by state
# normalized by the number of unique stations  

In [89]:
# Let's visualize the annual traffic volume by state
fips_state_abb = mappings.get("fips_state_abb")
fips_state_full = mappings.get("fips_state_full")

merge_station_state["fips_state_abb"] = merge_station_state.fips_state_code.apply(lambda x: fips_state_abb[x])
merge_station_state["fips_state_full"] = merge_station_state.fips_state_code.apply(lambda x: fips_state_full[x])

merge_station_state["norm_annual_traffic_volume"] = merge_station_state.annual_traffic_volume / merge_station_state.num_unique_stations

merge_station_state

fips_state_code  num_unique_stations  annual_traffic_volume fips_state_abb  \
0               01                  296             1461509905             AL   
1               02                  210              318138924             AK   
2               04                  526             5600393206             AZ   
3               05                  115              278436058             AR   
4               06                  570             6670057776             CA   
5               08                  212              964145461             CO   
6               09                   75              500556105             CT   
7               10                  168              534104309             DE   
8               11                    4               59205437             DC   
9               12                  545             4412694931             FL   
10              13                  446             3094751035             GA   
11              15                  152             1605036832             HI   
12              16                  423             1134837654             ID   
13              17                  205              604663105             IL   
14              18                  171              332236497             IN   
15              19                  201              510570289             IA   
16              20                  196              340832204             KS   
17              21                  162              597736449             KY   
18              22                  120              263211358             LA   
19              23                  129              195938846             ME   
20              24                  150             1399825682             MD   
21              25                  103              795943829             MA   
22              26                  245             2308504217             MI   
23              27                  177              828942796             MN   
24              28                  194              619803351             MS   
25              29                  168              849276076             MO   
26              30                  197              255608184             MT   
27              31                   99              329863440             NE   
28              32                  196             1037072029             NV   
29              33                  195             1108253743             NH   
30              34                   75              549456912             NJ   
31              35                  228              384977104             NM   
32              36                  345             1560213283             NY   
33              37                  134              466053858             NC   
34              38                  128               80318777             ND   
35              39                  352             2631835767             OH   
36              40                  180             1090108811             OK   
37              41                  350             1443103445             OR   
38              42                  180              534070530             PA   
39              44                  184             2441569023             RI   
40              45                  131             1511178572             SC   
41              46                  131              147911499             SD   
42              47                  114              276729299             TN   
43              48                  528             2791691757             TX   
44              49                  220             1191135280             UT   
45              50                  230              197420518             VT   
46              51                  961             9056241744             VA   
47              53                  327             2702688901             WA   
48              54                

In [99]:
fig = go.Figure()

fig.add_trace(go.Choropleth(
    locations=merge_station_state.fips_state_abb,
    z=merge_station_state["annual_traffic_volume"],
    locationmode="USA-states",
    colorscale="Reds",
    autocolorscale=False,
    colorbar_title="Billions",
    text=merge_station_state.fips_state_full,
    marker_line_color="white",
))

fig.update_layout(
    title="2015 US Traffic Volume by State",
    geo = dict(
        scope="usa", # Limit map scope to USA
        projection=go.layout.geo.Projection(type = 'albers usa'),
        showlakes=True, # lakes
        lakecolor="rgb(255, 255, 255)"),
    )

In [97]:
merge_station_state.sort_values("annual_traffic_volume", ascending=False).head()

fips_state_code  num_unique_stations  annual_traffic_volume fips_state_abb  \
46              51                  961             9056241744             VA   
4               06                  570             6670057776             CA   
2               04                  526             5600393206             AZ   
9               12                  545             4412694931             FL   
10              13                  446             3094751035             GA   

   fips_state_full  norm_annual_traffic_volume  
46        Virginia                9.423769e+06  
4       California                1.170186e+07  
2          Arizona                1.064714e+07  
9          Florida                8.096688e+06  
10         Georgia                6.938904e+06

In [98]:
merge_station_state.sort_values("norm_annual_traffic_volume", ascending=False).head()

fips_state_code  num_unique_stations  annual_traffic_volume fips_state_abb  \
8               11                    4               59205437             DC   
39              44                  184             2441569023             RI   
4               06                  570             6670057776             CA   
40              45                  131             1511178572             SC   
2               04                  526             5600393206             AZ   

         fips_state_full  norm_annual_traffic_volume  
8   District of Columbia                1.480136e+07  
39          Rhode Island                1.326940e+07  
4             California                1.170186e+07  
40        South Carolina                1.153571e+07  
2                Arizona                1.064714e+07

In [101]:
fig = go.Figure()

fig.add_trace(go.Choropleth(
    locations=merge_station_state.fips_state_abb,
    z=merge_station_state["norm_annual_traffic_volume"],
    locationmode="USA-states",
    colorscale="Reds",
    autocolorscale=False,
    colorbar_title="Billions",
    text=merge_station_state.fips_state_full,
    marker_line_color="white",
))

fig.update_layout(
    title="2015 US Traffic Volume by State",
    geo = dict(
        scope="usa", # Limit map scope to USA
        projection=go.layout.geo.Projection(type = 'albers usa'),
        showlakes=True, # lakes
        lakecolor="rgb(255, 255, 255)"),
    )